daemon/containerd/image_exporter.go

```diff
@@ -58,49 +59,93 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
 		archive.WithPlatform(platform),
 	}
 
-	ctx, release, err := i.client.WithLease(ctx)
-	if err != nil {
-		return errdefs.System(err)
-	}
-	defer release(ctx)
-
 	for _, name := range names {
 		target, err := i.resolveDescriptor(ctx, name)
 		if err != nil {
 			return err
 		}
 
-		// We may not have locally all the platforms that are specified in the index.
-		// Export only those manifests that we have.
-		// TODO(vvoland): Reconsider this when `--platform` is added.
-		if containerdimages.IsIndexType(target.MediaType) {
-			desc, err := i.getBestDescriptorForExport(ctx, target)
-			if err != nil {
-				return err
+		var cleanup func() error
+		if cleanup, err = i.leaseContent(ctx, target, func() error {
+			// We may not have locally all the platforms that are specified in the index.
+			// Export only those manifests that we have.
+			// TODO(vvoland): Reconsider this when `--platform` is added.
+			if containerdimages.IsIndexType(target.MediaType) {
+				desc, err := i.getBestDescriptorForExport(ctx, target)
+				if err != nil {
+					return err
+				}
+				target = desc
 			}
-			target = desc
-		}
 
-		if ref, err := reference.ParseNormalizedNamed(name); err == nil {
-			ref = reference.TagNameOnly(ref)
-			opts = append(opts, archive.WithManifest(target, ref.String()))
+			if ref, err := reference.ParseNormalizedNamed(name); err == nil {
+				ref = reference.TagNameOnly(ref)
+				opts = append(opts, archive.WithManifest(target, ref.String()))
 
-			log.G(ctx).WithFields(logrus.Fields{
-				"target": target,
-				"name":   ref.String(),
-			}).Debug("export image")
-		} else {
-			opts = append(opts, archive.WithManifest(target))
+				log.G(ctx).WithFields(logrus.Fields{
+					"target": target,
+					"name":   ref.String(),
+				}).Debug("export image")
+			} else {
+				opts = append(opts, archive.WithManifest(target))
 
-			log.G(ctx).WithFields(logrus.Fields{
-				"target": target,
-			}).Debug("export image without name")
+				log.G(ctx).WithFields(logrus.Fields{
+					"target": target,
+				}).Debug("export image without name")
+			}
+			return nil
+		}); err != nil {
+			cleanup()
```

Comment:
do we need to do anything with the error returned by cleanup? if we don't handle the error, we could remove it from the signature perhaps. (and if the errors are worth logging, the `cleanup()` returned could perhaps handle logging?

Reply:
I removed the error returned by cleanup and added your proposed changes :+1: